indexer_app/utils.py

```diff
@@ -876,6 +908,119 @@ async def handle_new_donation(
+    reg = await NadabotRegistry.objects.filter(id=receiverId).aupdate(
+        **{"default_human_threshold": data["default_human_threshold"]}
+    )
+
+    logger.info("updated threshold..")
+
+
+async def handle_new_provider(
+        data: dict,
+        receiverId: str,
+        signerId: str
+):
+    logger.info(f"new provider data: {data}, {receiverId}")
+    data = data["provider"]
+
+    logger.info(
+        f"upserting accounts involved, {data['submitted_by']}"
+    )
+
+    try:
+        submitter, _ = await Account.objects.aget_or_create(id=data["submitted_by"])
+
+        provider_id = data["id"]
+
+        if provider_id == 13:
+                provider_id = await cache.aget("last_id", 1)
+                await cache.aset("last_id", provider_id+1)
+
+        provider = await Provider.objects.aupdate_or_create(
+                id=provider_id,
+                contract_id=data["contract_id"],
+                method_name=data["method_name"],
+                provider_name=data["provider_name"],
+                description=data.get("description"),
+                status=data["status"],
+                admin_notes=data.get("admin_notes"),
+                default_weight=data["default_weight"],
+                gas=data.get("gas"),
+                tags=data.get("tags"),
+                icon_url=data.get("icon_url"),
+                external_url=data.get("external_url"),
+                submitted_by_id=data["submitted_by"],
+                submitted_at_ms = datetime.fromtimestamp(data.get("submitted_at_ms") / 1000),
+                stamp_validity_ms = datetime.fromtimestamp(data.get("stamp_validity_ms") / 1000) if data.get("stamp_validity_ms") else None,
+                account_id_arg_name = data["account_id_arg_name"],
+                custom_args = data.get("custom_args"),
+                registry_id=receiverId
+        )
+    except Exception as e:
+        logger.error(f"Failed to add new stamp provider: {e}")
+
+
+async def handle_add_stamp(
+        data: dict,
+        receiverId: str,
+        signerId: str
+):
+    logger.info(f"new stamp data: {data}, {receiverId}")
+    data = data["stamp"]
+
+    logger.info(f"upserting accounts involved, {data['user_id']}")
+
+    user, _ = await Account.objects.aget_or_create(id=data["user_id"])
+
+    try:
+        stamp = await Stamp.objects.aupdate_or_create(
+            user=user,
+            provider_id=data["provider_id"],
+            verification_date = datetime.fromtimestamp(data["validated_at_ms"] / 1000)
+        )
+    except Exception as e:
+        logger.error(f"Failed to create stamp: {e}")
+
+
+
+async def handle_new_group(
+    data: dict,
+    created_at: datetime
+):
+    logger.info(f"new group data: {data}")
+    group_data = data.get('group', {})
+    try:
+        rule = group_data['rule']
+        rule_key = rule
+        rule_val = None
+        if type(rule) == dict:
+            rule_key = next(iter(rule))
+            rule_val = rule.get(rule_key)
```

Comment:
since group enums can have values, they are represented as a dict in the events from the indexer, and enum choices without values are presented as normal strings:
withValue: `{'group': {'id': 5, 'name': 'Here we go again', 'providers': [8, 1, 4, 6], 'rule': {'IncreasingReturns': 10}}}`
noValue:  `{"id":6,"name":"Lachlan test group","providers":[1,2],"rule":"Highest"}`

Reply:
thanks, can you add this as a comment in the code please